keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,969 @@
+The following policy resources also exist in the cluster:
+- (1) A ClusterNetworkPolicy `Deny` rule selects [x/a] and denies all ingress traffic from Namespace y.
+- (2) A ClusterNetworkPolicy `Allow` rule selects [x/a, x/b] and allows all ingress traffic Namespace y.
+- (3) A NetworkPolicy rule isolates [x/b], only allows ingress traffic from Namespace z.
+- (4) A NetworkPolicy rule isolates [x/c], only allows ingress traffic from Namespace y.
+- (5) A DefaultNetworkPolicy rule isolates [x/c, x/d], only allows ingress traffic from Namespace z.
+
+Now suppose Pod y/client initiates traffic towards x/a, x/b, x/c and x/d.
+- y/client -> x/a is affected by rule (1) and (2). Since rule (1) has higher precedence, the request should be denied.
+- y/client -> x/b is affected by rule (2) and (3). Since rule (2) has higher precedence, the request should be allowed.
+- y/client -> x/c is affected by rule (4) and (5). Since rule (4) has higher precedence, the request should be allowed.
+- y/client -> x/d is affected by rule (5) only, The request should be denied.
+
+### User Stories
+
+![Alt text](user_story_diagram.png?raw=true "User Story Diagram")
+
+#### Story 1: Deny traffic from certain sources
+
+As a cluster admin, I want to explicitly deny traffic from certain source IPs
+that I know to be bad.
+
+Many admins maintain lists of IPs that are known to be bad actors, especially
+to curb DoS attacks. A cluster admin could use ClusterNetworkPolicy to codify
+all the source IPs that should be denied in order to prevent that traffic from
+accidentally reaching workloads. Note that the inverse of this (allow traffic
+from well known source IPs) is also a valid use case.
+
+#### Story 2: Funnel traffic through ingress/egress gateways
+
+As a cluster admin, I want to ensure that all traffic coming into (going out of)
+my cluster always goes through my ingress (egress) gateway.
+
+It is common practice in enterprises to setup checkpoints in their clusters at
+ingress/egress. These checkpoints usually perform advanced checks such as
+firewalling, authentication, packet/connection logging, etc.
+This is a big request for compliance reasons, and ClusterNetworkPolicy can ensure
+that all the traffic is forced to go through ingress/egress gateways.
+It is worth noting that the Cluster-scoped NetworkPolicy APIs will not redirect
+traffic, rather it can ensure that no traffic is allowed in/out except traffic
+via the gateways.
+
+#### Story 3: Isolate multiple tenants in a cluster
+
+As a cluster admin, I want to isolate all the tenants (modeled as Namespaces)
+on my cluster from each other by default.
+
+Many enterprises are creating shared Kubernetes clusters that are managed by a
+centralized platform team. Each internal team that wants to run their workloads
+gets assigned a Namespace on the shared clusters. Naturally, the platform team
+will want to make sure that, by default, all intra-namespace traffic is allowed
```

Comment:
You are correct in that, to enforce such tenant isolation, we needed 'except' rules to counter the 'deny' rules. I have added a new 'authorize' action in the latest commit to capture this. 

This story, however, was set out to express a different intent: tenant isolation is _not_ enforced, but rather provided as a starting point (or guideline) for initial security postures in tenants. Tenants can actually completely overwrite this based on the need of its namespaces. Of course, they cannot overwrite `Allow`, `Deny` and `Authorize` rules which cluster admins listed out as guardrails (dns must be allowed, egress to some IPs must be denied etc.)

If the cluster admin wants to _enforce_ tenant isolation instead, then it's a complete different story. As a matter of fact, this becomes Story 4 below.

Reply:
I'm not sure it's covered by story 4. Story 4 talks about allowing cluster-wide which seems very different to the use case of enforcing strict boundaries across namespaces which cannot be broken by the tenants themselves. This is one of the most common use cases we see Cilium users using cluster-wide policies for.

Reply:
If you look at the updated sample specs for story 4, I have actually written out how a strict boundary across namespace can be enforced:
```yaml
apiVersion: networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
spec:
  appliedTo:
    namespaceSelector:
      matchLabels:
        type: tenant  # assuming all tenant namespaces will be created with this label
  ingress:
    - action: Authorize
      from:
      - namespaces:
          self: true
    - action: Deny
      from:
      - namespaces:
          selector: {}
```

Reply:
I still think that this story doesn't truly justify a DNP, it only justifies existence because CNPs in the current proposal don't  enable a default-deny behavior.

If we aligned CNP to act more like existing network policies in that they apply a default-deny to selected pods, then this story no longer needs a new resource and can instead be implemented entirely with ClusterNetworkPolciy and the existing NetworkPolicy API.

e.g.,

- A single CNP that selects all pods, allows nothing (enabling default deny behavior).
- A NP per-namespace that selects all pods, and allows to/from all pods within the namespace.

Reply:
^ I think that has the bonus win of making CNP and NP behave more intuitively. The fact that one has a default deny and the other doesn't feels like a real area for confusion.

Reply:
```
- action: Deny 
  from: 
   - namespaces: 
        notSame: 
        - key: kubernetes.io/metadata.name
```
this sounds like neat alternative that we have not explored to me. I'll put more thought into it.

Reply:
Are there any use cases for "same"/"not-same" with CNPs?

It seems like the set of things people want to do with "allow/deny X by default" rules ("DNPs") doesn't really have a _huge_ overlap with the set of things people want to do with "forcibly allow/deny X" rules ("CNPs"), so maybe they shouldn't actually be so similar?

Like, there is little point in creating a DNP that says "all egress is blocked by default, except to CoreDNS", because as soon as the user creates their own NetworkPolicy to allow the egress they need, they have to remember to re-allow CoreDNS again, because their NP would override the DNP. It would be much simpler for everyone if the admin just created a DNP saying "all egress is blocked by default, period" and a CNP saying "egress to CoreDNS is allowed, always".

So, if CNPs don't need "same-namespace" and DNPs don't need... lots of things, then maybe DNP shouldn't be like CNP, but should be more tightly focused on just establishing the baseline isolation/tenancy model, using a set of fields/selectors that make sense for the baseline isolation/tenancy use cases.

Reply:
I think same/notSame in CNP enables multi-namespace tenancy in a way nothing else does.  For hard tenancy, you really don't want it to be user-overrideable.

IOW - a user in a tenant=coke namespace is allowed to talk to other tenant=coke namespace OR to tenant=shared; but they still have to use network policy to do so.  And they can talk to DNS

CNP = [ deny * except [ notSame(tenant), tenant=shared ] ; allow DNS ]
DNP = allow nothing
NP = <explicit>

The CNP expression is what worries me.  I don't want to "allow same(tenant)" because that is a unilateral allow (including things that user NP doesn't want).  I want to deny everything except same tenant and system stuff, both of which are "allowable" but not "allowed".  Can we express that more succinctly?

something like:

CNP = [ empower same(tenant); empower tenant=system; allow DNS ]

that seems more expressive - empowered != allowed